overlord/snapstate/snapstate.go

```diff
@@ -2250,6 +2250,53 @@ func AddLinkNewBaseOrKernel(st *state.State, ts *state.TaskSet) (*state.TaskSet,
 	return ts, nil
 }
 
+func SwitchToNewGadget(st *state.State, name string) (*state.TaskSet, error) {
+	var snapst SnapState
+	err := Get(st, name, &snapst)
+	if err == state.ErrNoState {
+		return nil, &snap.NotInstalledError{Snap: name}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if err := CheckChangeConflict(st, name, nil); err != nil {
+		return nil, err
+	}
+
+	info, err := snapst.CurrentInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	if info.Type() != snap.TypeGadget {
+		return nil, fmt.Errorf("cannot link type %v", info.Type())
```

Comment:
shouldn't this be an internal error if we accidentally pass in the name of a non-gadget snap to SwitchToNewGadget? in other words, shouldn't we have caught that specific error earlier somewhere?